### DISCUSSION TITLE
AWS ECS Fargate and Application Load Balancer

### DISCUSSION THREAD

**Discussion by fwhenin** [keycloak/keycloak#12028] [Q&A]
Hello, I'm looking for a bit of help in our Keycloak Deployment

I decided to put Keycloak on ECS Fargate with an Application Load Balancer in front of it, I reached the point where I'm able to access the welcome page. I even have a custom subdomain in Route 53 (login.domain.com) pointing to the Load Balancer. I even have it connected to Rds with a Postgresql instance

The issue I'm running into is that when I hover/click on the "Administration Console" the url is "https://0.0.0.0/admin/", not "https://login.domain.com/admin/" as i would expect. when I go to "https://login.domain.com/realms/master/.well-known/openid-configuration" almost everything looks right except "token_endpoint", "introspecting_endpoint", "userinfo_endpoint" and "jwks_uri" all has 0.0.0.0 in the url and not the custom domain one

`Here are my Environment Variables for ECS
        KC_PROXY: "edge",
        KC_HEALTH_ENABLED: "true",
        KC_HOSTNAME: `login.domain.com`,
`

I tried setting KC_HTTP_HOST, and Keycloak was not successfully spinning up then. Any help would be much appreciated as this has flustered me all day today

**Comment by DGuhr** [comment]
a guess for now: When you want all endpoints to run over the publicly available hostname, you want to set `KC_HOSTNAME_STRICT_BACKCHANNEL=true`. Could you try that and see if it works for you?

ref: https://www.keycloak.org/server/hostname#_backend_endpoints

**Comment by fwhenin** [comment]
@DGuhr OK, thanks for that tip, that partially worked though, in the openid-configuration endpoint, I'm not seeing ANY 0.0.0.0. 

But when I go to the admin ui at login.domain.com/admin, it still forwards me to 0.0.0.0/admin. I have the KC_HOSTNAME env variable set, is there something else I'm missing? does KC_HTTP_HOST need set too?

**Reply by DGuhr** [reply]
Hum, no idea tbh. @pedroigor perhaps?

**Reply by fwhenin** [reply]
Any ideas what happens if X-Forwarded-Host isn't present? just realized that AWS Application Load Balancer [doesn't support that header](https://docs.aws.amazon.com/elasticloadbalancing/latest/application/x-forwarded-headers.html)

**Reply by pedroigor** [reply]
Perhaps this is the issue. If the host is not set the admin URLs might be created based on the address the proxy is accessing the server.

I think we need to bring back the option to force a hostname for the admin if that is what is happening.

**Reply by pedroigor** [reply]
@fwhenin This should help https://github.com/keycloak/keycloak/pull/12192.

**Comment by fwhenin** [comment]
@DGuhr I went back and used the 17.0.0 version, and I'm not seeing the issue i'm having anymore, this seems to be an issue with 18.0. I'm about to try 17.0.1 to see if it's happening there

**Comment by wadahiro** [comment]
@fwhenin @pedroigor 

Hello,

I also recently encountered same issue with AWS ECS Fargate and Application Load Balancer (ALB) in `KC_PROXY=edge` mode, resulting in https://0.0.0.0/admin/. The Keycloak version is the latest at the moment, 21.1.1. I have found the cause of the problem and hope it helps someone else.

The procedure to reproduce the problem is to set the `ALB --> Keycloak` access to **HTTP/2**, which reproduces it 100%.

ALB does not send `X-Forwarded-Host` due to its specification [1]. Instead, `Host` header is sent, which is fine when using `HTTP/1.1`. Keycloak (Quarkus) looks at the `Host` header and works as expected.

However, this problem occurs when using **HTTP/2** with the functionality in the ALB documentation [2].

After getting the TCP dump and investigating the cause, ALB was sending the `Host` header even for `HTTP/2`. However, it did not send the `:authority` header, which is recommended to be used in the HTTP/2 specification [3]. This unfortunately causes problems with Keycloak (Quarkus), which seems to look at the **:authority** header when processing HTTP/2 request.

I contacted AWS support about this and they said that although it is not ALB's specification to not put the `:authority` header, it is the current behaviour.

It would be nice if the Keycloak (Quarkus) side could handle the no `:authority` header situation (I will create a ticket), but until then, the workaround is to unfortunately not use HTTP/2.

- [1] https://docs.aws.amazon.com/elasticloadbalancing/latest/application/x-forwarded-headers.html
- [2] https://docs.aws.amazon.com/elasticloadbalancing/latest/application/load-balancer-target-groups.html#target-group-protocol-VERSION
- [3] https://httpwg.org/specs/rfc7540.html#rfc.section.8.1.2.3

**Reply by wadahiro** [reply]
I've created for this issue: https://github.com/keycloak/keycloak/issues/20959
